TrafficCapture/nettyWireLogging/src/test/java/org/opensearch/migrations/trafficcapture/netty/RootWireLoggingContextTest.java

```diff
@@ -0,0 +1,103 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static org.opensearch.migrations.trafficcapture.netty.TestStreamManager.consumeIntoArray;
+
+@Slf4j
+public class RootWireLoggingContextTest {
+
+  private static void writeMessageAndVerifyTraces(byte[] fullTrafficBytes, Consumer<EmbeddedChannel> channelWriter, boolean shouldBlock,
+                                                  List<String> expectedTraces) throws IOException {
+    try (var rootContext = new TestRootContext(true, true)) {
+      var streamManager = new TestStreamManager();
+      var offloader = new StreamChannelConnectionCaptureSerializer("Test", "c", streamManager);
+
+      EmbeddedChannel channel = new EmbeddedChannel(
+          new ConditionallyReliableLoggingHttpHandler(rootContext,
+              "n", "c", ctx -> offloader,
+              new RequestCapturePredicate(), x -> shouldBlock));
+      channelWriter.accept(channel);
+
+      // we wrote the correct data to the downstream handler/channel
+      var outputData = new SequenceInputStream(Collections.enumeration(channel.inboundMessages().stream()
+          .map(m -> new ByteArrayInputStream(consumeIntoArray((ByteBuf) m)))
+          .collect(Collectors.toList())))
+          .readAllBytes();
+      Assertions.assertArrayEquals(fullTrafficBytes, outputData);
+
+      // For a non-blocking write, close the channel to force it to flush
+      if (!shouldBlock) {
+        channel.close().awaitUninterruptibly();
+      }
+
+      Assertions.assertNotNull(streamManager.byteBufferAtomicReference.get(),
+          "This would be null if the handler didn't block until the output was written");
+
+      var trafficStream = TrafficStream.parseFrom(streamManager.byteBufferAtomicReference.get());
+      Assertions.assertTrue(trafficStream.getSubStreamCount() > 0 &&
+          trafficStream.getSubStream(0).hasRead());
+      var combinedTrafficPacketsStream =
+          new SequenceInputStream(Collections.enumeration(trafficStream.getSubStreamList().stream()
+              .filter(TrafficObservation::hasRead)
+              .map(to -> new ByteArrayInputStream(to.getRead().getData().toByteArray()))
+              .collect(Collectors.toList())));
+      Assertions.assertArrayEquals(fullTrafficBytes, combinedTrafficPacketsStream.readAllBytes());
+      Assertions.assertEquals(1, streamManager.flushCount.get());
+
+      List<SpanData> finishedSpans = rootContext.instrumentationBundle.getFinishedSpans();
+      Assertions.assertTrue(!finishedSpans.isEmpty());
+      Assertions.assertTrue(!rootContext.instrumentationBundle.getFinishedMetrics().isEmpty());
+
+      List<String> finishedSpanNames = finishedSpans.stream().map(x -> x.getName()).collect(Collectors.toList());
+      expectedTraces.forEach(trace ->
+          Assertions.assertTrue(finishedSpanNames.contains(trace), "finishedSpans does not contain " + trace)
+      );
```

Comment:
You could collect finishedSpanNames into a set too (& 'var' instead of 'List<String>' would let you swap back and forth.
An easier explanation of when things go wrong for this test could be to make a big string with the expected span names sorted & to sort the finishedSpans in the same way.  Something like
```
testSpanExporter.getFinishedSpanItems().stream().map(SpanData::getName).sorted().collect(Collectors.joining("\n"))
```
Now when those two lists don't match up, the user will understand how extensive the mismatch was.  Also, the cognitive load to assert that the test is complete is lower.

Reply:
ah, okay, I like this. I was currently considering `expectedTraces` to pass if it was a subset of the actual traces, which is what led to the `blocked` note below. BUt it's probably clearer to just make the tests specify the entire list of expected traces.